realtek: do good to the MDIO driver

plappermaul opened the pull request:
This PR does some good to the MDIO driver that lives somewhere in the ethernet sources.

- Rename variables/defines with rtmdio_ prefix that already was started during kernel 6.6 refactoring
- Fix a long standing out-of-bounds read
- Decouple MDIO and ethernet data structures

systemcrash commented:
ping @svanheule

plappermaul commented:
@andrewjlamarche @Borromini @achterin Regardless of thumbs up and the ping, I hope people here can test and approve the PR. In the past this led to much shorter processing times. Thanks in advance.

achterin commented:
@plappermaul anything specific you want tested or just see if nothing broke? for me personally it would help if you could just write down some "instructions" at the start of the pr of what you want tested and i will gladly help with anything i can.

thanks for your continuous effort to move this target forward!

plappermaul commented:
Thanks for that offer. For now (and the following separation PRs) it will be sufficient to

- boot old/new image from ramfs
- compare bootlogs (should be the same)
- do simple network setup
- check if you can reach the network by connecting the different ports

Finally comment on what device you tested and what was the result. If ok approve the PR.

andrewjlamarche commented:
I would test but the only Realtek device I have is still not fully supported and acts funky, so my input would probably not be considered useful at this time (see https://github.com/openwrt/openwrt/pull/17137)

achterin commented:
Successfully tested on XGS1250-12. Bootlogs look identical and a simple network test was successful on all ports.

plappermaul commented:
Thanks. Just for simplicity and documentation. Can you head over to the commits page and use the revew/approve function.

![image](https://github.com/user-attachments/assets/2d271a55-083e-4e12-812e-3bf2e66e2d18)

vincele commented:
The code modifications look fairly straightforward. The only thing that caught my attention was the duplication of `id` & `family_id` in both structs. FWIW, LGTM.

plappermaul commented:
Thanks for the review. MDIO will get its own driver so duplication is expected. Can you approve as well?

vincele reviewed (approved):
LGTM

janh commented:
Tested on HPE 1920-8G (rtl838x), HPE 1920-48G (rtl839x), Zyxel XGS1010-12 (rtl930x). I didn't notice any difference, which I guess is a good sign.

plappermaul commented:
@janh Thank you and please approve.

Neustradamus commented:
To follow this PR :)

plappermaul commented:
Hi @svanheule, @dangowrt 

@cpackham-atlnz is working hard to get RTL930x stuff upstream. See https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=24e31e474769 

With this patch (and subsequent ones) I'm trying to ease future migration and to avoid repeating the realtek kernel 6.6 mess of last year. Any chances that we can align to baby steps like "will send small patches" & "will review & approve promptly" so that we can proceed?

vivivitus reviewed (approved):
Tested on a XGS1250-12 V2. The ports are working like before (the three 10GBit ports are non working like before).
I recognized one new line:
REALTEK RTL9300 SERDES rtl838x slave mii-0:3f: Detected internal RTL9300 Serdes

plappermaul committed:
realtek: add mdio prefix to defines

Inside the ethernet driver lives the mdio bus. It is not always clear
what belongs where. Prefix some leftovers from the kernel 6.6 refactor
to clearly state what belongs to the bus. Group all defines together
in one place. This commit has no functional changes.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18402
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

plappermaul committed:
realtek: reuse RTMDIO_MAX_SMI_BUS define

Although a dfine is used to set the maxiumum number of SMI
busses (=4) it is not used at all appropriate places in the code.
Replace hard coded constants with that define.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18402
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

plappermaul committed:
realtek: move private bus structure closer to the bus

Relocate the bus structure definition into the MDIO source code area
of the ethernet driver. So if the real bus driver is forked from the
rest of the code only one area needs to be removed. Rename it to make
clear it belongs to the bus.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18402
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

plappermaul committed:
realtek: resize mdio bus private arrays

These two arrays have been fixed to some sane size (= 64 ports). Now
that everything is in place reuse the global RTMDIO_MAX_PORT define.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18402
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

plappermaul committed:
realtek: harden MDIO driver

At least since 2022 there is a major bug in the MDIO driver that
produces out-of-bound reads and erratic behaviour during initialization.

- mdiobus_scan_bus_c22() scans the bus for 64 devices (PHY_MAX_ADDR)
- private bus structure only supports 57 entry arrays (MAX_PORTS)

All the bus/reader writer functions accept calls with addr>=57 and will
silently read beyond their limits. This can lead to ghost SERDES like
https://github.com/openwrt/openwrt/issues/18665#issuecomment-2846053813

Add proper boundary checks and end the functions with -ENODEV that is
the only accepted error code from the bus scan function.

Fixes: 0536c582e673aa292 ("realtek: Fix RTL931X Ethernet driver") etc ...

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18402
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

plappermaul committed:
realtek: decouple MDIO and ethernet devices

We are lucky to have a working realtek environment. But some things where mixed
heavily. To say it clear a bus is a bus and an ethernet is an ethernet. With
the new naming conventions and defines this becomes even more obvious.

Decouple it by moving the bus specific parts out of the ethernet device. To
make the code more readable rename bus_priv variables to priv and sort variable
definitions in inverse tree order (length descending) where it makes sense.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/18402
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

plappermaul commented:
@vivivitus the SERDES ghost comes from a long standing bug. See issue https://github.com/openwrt/openwrt/issues/18665#issuecomment-2846053813 for another occurence. 

New PR version is availabe that includes the bugfix in "[realtek: harden MDIO driver](https://github.com/openwrt/openwrt/pull/18402/commits/389e481cc5ae5a9c320d79470fe616fd7db7a7ca)"

For retesting - especially on RTL83xx - I advise to forward to current OpenWrt main that includes a RTL8214FC fix https://github.com/plappermaul/openwrt/commit/241343a2f2f0d3a22b5fd2f16eb0a578c76b4948

jwh reviewed (approved):
Tested on GS1900-48; no functional change, erroneous detection no longer shows up

dekarl commented:
Tested on HPE 1920-24G (rtl838x), works as before.

Small changes in the boot log (serial console works, too)

```diff
--- dmesg.pre-mdio.no-timestamp	2025-05-04 21:44:04.047484111 +0000
+++ dmesg.mdio.no-timestamp	2025-05-04 21:43:49.055656361 +0000
@@ -1,4 +1,4 @@
-Linux version 6.6.87 (builder@buildhost) (mips-openwrt-linux-musl-gcc (OpenWrt GCC 13.3.0 r29254-a06ed856bb) 13.3.0, GNU ld (GNU Binutils) 2.42) #0 Fri Apr 18 00:10:48 2025
+Linux version 6.6.87 (dekarl@openwrtlab) (mips-openwrt-linux-musl-gcc (OpenWrt GCC 14.2.0 r29342+89-6b1d0ce88c) 14.2.0, GNU ld (GNU Binutils) 2.42) #0 Fri May  2 19:02:54 2025
 RTL838X model is 83826800
 SoC Type: RTL8382
 printk: bootconsole [early0] enabled
@@ -55,12 +55,11 @@
 workingset: timestamp_bits=14 max_order=15 bucket_order=1
 squashfs: version 4.0 (2009/01/31) Phillip Lougher
 jffs2: version 2.2 (NAND) (SUMMARY) (ZLIB) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
-Block layer SCSI generic (bsg) driver version 0.4 loaded (major 252)
 pinctrl-single 1b000144.pinmux: 32 pins, size 4
 pinctrl-single 1b001000.pinmux: 32 pins, size 4
 pinctrl-single 1b00a000.pinmux: 32 pins, size 4
 pinctrl-single 1b00a0e0.pinmux: 32 pins, size 4
-Serial: 8250/16550 driver, 16 ports, IRQ sharing enabled
+Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
 printk: console [ttyS0] disabled
 18002000.uart: ttyS0 at MMIO 0x18002000 (irq = 31, base_baud = 12500000) is a 16550A
 printk: console [ttyS0] enabled
@@ -87,7 +86,7 @@
 1 squashfs-split partitions found on MTD device rootfs
 0x0000004f0000-0x000001cf0000 : "rootfs_data"
 0x000001ff0000-0x000002000000 : "factory"
-Probing RTL838X eth device pdev: 818bfe00, dev: 818bfe10
+Probing RTL838X eth device pdev: 818bc200, dev: 818bc210
 Found SoC ID: 8382: RTL8382, family 8380
 rtl8380_init_mac
 Using MAC 00005c8a38866adf
@@ -129,7 +127,7 @@
 Phy on MAC 0: 1cc981
 sfp sfp-2: module OEM              GBIC-SX-SFP-3    rev  sn A9714335         dc 100114
 rtl8214fc_media_set: port 26, set_fibre: 1
-Realtek RTL8218B (external) rtl838x slave mii-0:00: Expected external RTL8218B, found PHY-ID 6b23
+Realtek RTL8218B (external) rtl838x slave mii-0:00: Expected external RTL8218B, found PHY-ID 2
 Realtek RTL8218B (external): probe of rtl838x slave mii-0:00 failed with error -1
 hwmon hwmon1: temp1_input not attached to any thermal zone
 rtl8214fc_power_set: Powering off COPPER (port 26)
```

plappermaul commented:
@dekarl thanks for posting. Regarding the error messages "rtl838x slave mii-0:00: Expected external RTL8218B, found PHY-ID". This does not come from this PR. If this needs to be fixed open an issue.

stintel commented:
Finally building an image for my ECS4100 with this PR included to test. Just to make sure: this PR does not depend on any other outstanding PRs?

plappermaul commented:
Nope. All my PRs are independent from each other.

stintel commented:
Thanks @plappermaul for your work and thanks all testers/approvers, I have tested this successfully on my ECS4100 and pushed it to master.

plappermaul commented:
@stintel thank you.

GoetzGoerisch commented:
@stintel Thank you. Should this be backported to 24.10?

stintel commented:
> @stintel Thank you. Should this be backported to 24.10?

I'll let someone else decide that. I'm too busy with too many things to take on these kinds of decisions.

plappermaul commented:
How can I mark this one as closed/merged?

stintel commented:
> How can I mark this one as closed/merged?

For some reason it didn't happen automatically with the script.
